File: IAIso-v5.0/core/iaiso-ruby/lib/iaiso/middleware/anthropic.rb
# frozen_string_literal: true

require_relative "errors"
require_relative "../core/pressure_engine"
require_relative "../core/step_outcome"

module IAIso
  module Middleware
    # Anthropic Messages API middleware.
    module Anthropic
      # Response struct — minimal subset of fields we account against.
      Response = Data.define(:model, :input_tokens, :output_tokens, :content)

      # A single content block. `type` is "text" or "tool_use".
      ContentBlock = Data.define(:type)

      Options = Data.define(:raise_on_escalation) do
        def self.defaults = new(raise_on_escalation: false)
      end

      # Wraps a duck-typed `raw` client (any object responding to
      # `messages_create(params) -> Response`) so every call is
      # accounted against a BoundedExecution.
      class BoundedClient
        def initialize(raw:, execution:, options: Options.defaults)
          @raw = raw
          @execution = execution
          @options = options
        end

        def messages_create(params)
          pre = @execution.check
          raise LockedError if pre == IAIso::Core::StepOutcome::LOCKED
          if pre == IAIso::Core::StepOutcome::ESCALATED && @options.raise_on_escalation
            raise EscalationRaisedError
          end
          resp =
            begin
              @raw.messages_create(params)
            rescue StandardError => e
              raise ProviderError, e.message
            end
          tokens = resp.input_tokens.to_i + resp.output_tokens.to_i
          tool_calls = resp.content.count { |b| b.type == "tool_use" }
          model = resp.model.to_s.empty? ? "unknown" : resp.model
          @execution.record_step(IAIso::Core::StepInput.build(
            tokens: tokens, tool_calls: tool_calls,
            tag: "anthropic.messages.create:#{model}",
          ))
          resp
        end
      end
    end
  end
end

File: IAIso-v5.0/core/iaiso-rust/crates/middleware/src/anthropic.rs
//! Anthropic Messages API wrapper.

use crate::MiddlewareError;
use iaiso_core::{BoundedExecution, StepInput, StepOutcome};
use serde::Deserialize;
use serde_json::Value;

#[derive(Debug, Clone, Deserialize, Default)]
pub struct Usage {
    #[serde(default)]
    pub input_tokens: u64,
    #[serde(default)]
    pub output_tokens: u64,
}

#[derive(Debug, Clone, Deserialize)]
pub struct ContentBlock {
    #[serde(rename = "type")]
    pub kind: String,
}

#[derive(Debug, Clone, Deserialize, Default)]
pub struct Response {
    #[serde(default)]
    pub model: String,
    #[serde(default)]
    pub usage: Usage,
    #[serde(default)]
    pub content: Vec<ContentBlock>,
}

/// Structural interface a wrapped Anthropic client must satisfy.
pub trait Client: Send + Sync {
    fn messages_create(&self, params: &Value) -> Result<Response, String>;
}

#[derive(Debug, Clone, Copy, Default)]
pub struct Options {
    pub raise_on_escalation: bool,
}

pub struct BoundedClient<'a, C: Client> {
    raw: C,
    execution: &'a BoundedExecution,
    opts: Options,
}

impl<'a, C: Client> BoundedClient<'a, C> {
    pub fn new(raw: C, execution: &'a BoundedExecution, opts: Options) -> Self {
        Self { raw, execution, opts }
    }

    pub fn messages_create(&self, params: &Value) -> Result<Response, MiddlewareError> {
        match self.execution.check() {
            StepOutcome::Locked => return Err(MiddlewareError::Locked),
            StepOutcome::Escalated if self.opts.raise_on_escalation => {
                return Err(MiddlewareError::EscalationRaised)
            }
            _ => {}
        }
        let resp = self
            .raw
            .messages_create(params)
            .map_err(MiddlewareError::Provider)?;
        let tokens = resp.usage.input_tokens + resp.usage.output_tokens;
        let tool_calls = resp
            .content
            .iter()
            .filter(|b| b.kind == "tool_use")
            .count() as u64;
        let model = if resp.model.is_empty() {
            "unknown".to_string()
        } else {
            resp.model.clone()
        };
        let _ = self.execution.record_step(StepInput {
            tokens,
            tool_calls,
            tag: Some(format!("anthropic.messages.create:{}", model)),
            ..Default::default()
        });
        Ok(resp)
    }
}
